### PR TITLE
Updates to build for JEDI linking/control, add wcoss2

### LINKED ISSUE #294: Allow JEDI control of UFS via NUOPC/ESMF
## Description
Some changes are required within UFS to allow for JEDI to link to UFS components and control the clock and state variables for data assimilation.

## Solution
Create individual libraries with exposed dependencies during the UFS build that can be linked into the JEDI data assimilation system. Set up build on NEMS side to allow for control from outside of NEMS (NEMS is no longer the top level driver). 

## Alternatives
So far, these changes are related to the build system, but there may be further modifications to source code required to enable JEDI to rewind the clock for the UFS and restart with updated state variables.

## Related to (coming shortly)
Directly reference any issues or PRs in this or other repositories that this is related to, and describe how they are related. Example:
- required to support noaa-emc/nems/issues/<issue_number>
- needed by noaa-emc/fv3atm/pull/<pr_number>

### PR DESCRIPTION
**Pull request by mark-a-potts**
## Description

This change updates the build system to allow JEDI to find and link to libufs, libfv3atm, and libfms. This change also incorporates an optional build directive to allow an ESMF/NUOPC driver above NEMS. The optional directive is called NEMS_DRIVER and defaults to ON to keep behavior as currently expected. Adding -DNEMS_DRIVER=OFF to the cmake configure command will allow JEDI or another ESMF/NUOPC code to manipulate the clock of the UFS. 

### Issue(s) addressed

fixes ufs-community/ufs-weather-model/issues/294

## Testing

Testing is currently underway on Hera
This only really affects the build system, so no changes to results are expected.

## Dependencies

- waiting on noaa-emc/nems/pull/85
- waiting on noaa-emc/fv3atm/pull/211 (includes noaa-emc/fv3atm/pull/201)

### CODE REVIEW

File: CMakeLists.txt
Comment (junwang-noaa) [inline CMakeLists.txt:157]: May I ask which FMS version will be used in this PR? My understand is that we are moving to FMS 2020.04 in early Dec, currently using fms2_io may cause issues in MOM6.

File: CMakeLists.txt
Comment (junwang-noaa) [inline CMakeLists.txt]: We'd like to have NEMS_DRIVER as default, and use JEDI_DRIVER if the model needs to be run within JEDI.

**Review by junwang-noaa — COMMENTED** [review]
Please check your fv3 component, it seems you point to a revision: 45f1be37d , which shows on noaa -emc fv3 repo as: This commit does not belong to any branch on this respository, and may belong to a fork outside of the repository.

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt:157]: The regression tests were run with 2019.01.03, but the changes should work with newer versions as well. I had earlier tested with 2020.02, so I think that is why the fms2_io line is in there. I can remove it for this PR.

File: CMakeLists.txt
Comment (mark-a-potts) [reply CMakeLists.txt]: I'll make the changes to the nomenclature.

**Review by climbfuji — CHANGES_REQUESTED** [review]
I am missing the temporary updates to `.gitmodules`. We need to be able to check out the code recursively (git clone -b jedi-ufs --recursive https://github.com/mark-a-potts/ufs-weather-model and get exactly the code that needs to be tested, i.e. including all changes in fv3atm and NEMS). Something is wrong with the fv3atm PR, by the way, it is showing a large number of changes that have been merged previously.

EDIT: I can check out the code recursively, even though `.gitmodules` does not contain the updated URLs. Will send a PR for fv3atm your way to fix https://github.com/NOAA-EMC/fv3atm/pull/201.

**Comment by climbfuji** [comment]
The last commits added the regression test logs twice, once where they belong (in `tests/`), and then again in the top-level directory. Can you remove the latter, please?

**Comment by climbfuji** [comment]
@mark-a-potts please convert this PR to "ready for review" and remove the extraneous regression test log files in the top-level directory. Thank you!

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by DusanJovic-NOAA** [comment]
This PR depend on https://github.com/NOAA-EMC/fv3atm/pull/211 not https://github.com/NOAA-EMC/fv3atm/pull/201

**Comment by climbfuji** [comment]
> This PR depend on [NOAA-EMC/fv3atm#211](https://github.com/NOAA-EMC/fv3atm/pull/211) not [NOAA-EMC/fv3atm#201](https://github.com/NOAA-EMC/fv3atm/pull/201)

This is my mistake, I didn't update the description when 211 replaced 201.

**Review by climbfuji — CHANGES_REQUESTED** [review]
Please remove the following four files from the top-level directory
```
RegressionTests_cheyenne.gnu.log
RegressionTests_cheyenne.intel.log
RegressionTests_gaea.intel.log
RegressionTests_jet.intel.log
```
and add the missing
```
tests/RegressionTests_wcoss_dell_p3.log
```

File: .gitmodules
Comment (climbfuji) [inline .gitmodules]: Please revert this change.

**Review by climbfuji — COMMENTED** [review]
Please update the fv3atm submodule pointer to point to the head of noaa-emc develop (hash is babf6dc3e904ea9fd4e9773bf792ada072e4a427) - and same for NEMS once that one gets merged - and revert the change to `.gitmodules`. Thank you!

**Review by junwang-noaa — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]

**Comment by junwang-noaa** [comment]
I will wait for the CI test to finish, the do the commit.

**Comment by climbfuji** [comment]
@junwang-noaa All tests passed, you can press the magic button anytime. Happy Weekend!

**Pull request merged**
